### PR TITLE
Volumes density lookup

### PR DESCRIPTION
**Pull request by V-Gorash**
JIRA TICKET: RPRMAYA-2895, RPRMAYA-2878

EFFECT OF CHANGES: 
        VDB volumes: density multiplier enabled, random crash fix
        Maya volumes: density displays correctly

NOTE: some scenes with clouds still is being rendered incorrectly, but this issue can be reproduced with RPR 1. The cause of this issue is 3D textures from Maya which we can't get. Arnold use undocumented way to get these textures

### CODE REVIEW

**Review by Vsevolod1983 — APPROVED** [review]

File: FireRender.Maya.Src/FireRenderVolume.cpp [hunk 1]
```diff
@@ -430,7 +430,7 @@ bool NorthstarRPRVolume::TranslateVolume()
 	m_emissionGrid.Reset();
 
 
-	if (vdata.densityGrid.IsValid()) // grid exists
+	if (vdata.densityGrid.IsValid() && vdata.densityGrid.gridOnIndices.size() > 0) // grid exists
```
Comment (DMotorygin) [inline FireRender.Maya.Src/FireRenderVolume.cpp:433]: Please consider moving check gridOnIndices.size() > 0 to IsValid() method.

If I understood correctly grid with gridOnIndices.size() == 0 would be invalid, right?

File: FireRender.Maya.Src/FireRenderVolume.cpp [hunk 2]
```diff
@@ -441,19 +441,16 @@ bool NorthstarRPRVolume::TranslateVolume()
 			RPR_GRID_INDICES_TOPOLOGY_XYZ_U32
 		);
 
+		MFnDependencyNode depNode(node);
+		float maxDensityValue = RPRVolumeAttributes::GetDensityMultiplier(depNode) * 1.1f;
```
Comment (DMotorygin) [inline FireRender.Maya.Src/FireRenderVolume.cpp:445]: Please make 1.1f multipler a constant and add a comment where 1.1 value comes from

File: FireRender.Maya.Src/FireRenderVolume.cpp [hunk 3]
```diff
@@ -441,19 +441,16 @@ bool NorthstarRPRVolume::TranslateVolume()
...
 		volumeShader.xSetValue(RPR_MATERIAL_INPUT_DENSITYGRID, densityGridNode);
-
-		//now density is a constant value because density lookup is temporary unavailable in Northstar
-		//volumeShader.xSetValue(RPR_MATERIAL_INPUT_DENSITY, densityLookupNode);
-		volumeShader.xSetParameterF(RPR_MATERIAL_INPUT_DENSITY, 1000.0f, 1.0f, 1.0f, 1.0f); 
+		volumeShader.xSetParameterF(RPR_MATERIAL_INPUT_DENSITY, maxDensityValue, 1.0f, 1.0f, 1.0f);
 	}
 	
-	if (vdata.albedoGrid.IsValid()) // grid exists
+	if (vdata.albedoGrid.IsValid() && vdata.albedoGrid.gridOnIndices.size() > 0) // grid exists
```
Comment (DMotorygin) [inline FireRender.Maya.Src/FireRenderVolume.cpp:453]: Please consider moving check gridOnIndices.size() > 0 to IsValid() method.

If I understood correctly grid with gridOnIndices.size() == 0 would be invalid, right?

File: FireRender.Maya.Src/FireRenderVolume.cpp [hunk 4]
```diff
@@ -471,7 +468,7 @@ bool NorthstarRPRVolume::TranslateVolume()
 		volumeShader.xSetValue(RPR_MATERIAL_INPUT_COLOR, albedoLookupNode);
 	}
 
-	if (vdata.emissionGrid.IsValid()) // grid exists
+	if (vdata.emissionGrid.IsValid() && vdata.emissionGrid.gridOnIndices.size() > 0) // grid exists
```
Comment (DMotorygin) [inline FireRender.Maya.Src/FireRenderVolume.cpp:471]: Please consider moving check gridOnIndices.size() > 0 to IsValid() method.

If I understood correctly grid with gridOnIndices.size() == 0 would be invalid, right?

File: FireRender.Maya.Src/FireRenderVolume.cpp [hunk 5]
```diff
@@ -1735,8 +1755,7 @@ bool NorthstarFluidVolume::TranslateVolume(void)
 	
 	// apply nodes to main shader
 	volumeShader.xSetValue(RPR_MATERIAL_INPUT_DENSITYGRID, densityGridNode);
-	//volumeShader.xSetValue(RPR_MATERIAL_INPUT_DENSITY, densityLookupNode); // not implemented in northstar
-	volumeShader.xSetParameterF(RPR_MATERIAL_INPUT_DENSITY, 10000.f, 1.f, 1.f, 1.f); // constant mock
+	volumeShader.xSetParameterF(RPR_MATERIAL_INPUT_DENSITY, 1000.f, 1.f, 1.f, 1.f); // multiplication coefficient
```
Comment (DMotorygin) [inline FireRender.Maya.Src/FireRenderVolume.cpp:1758]: Why do we still need a 1000 constant multiplier if we modify grid values directly?

If yes, it was 10 000 in previous version. Was this change intentional?

**Review by DMotorygin — CHANGES_REQUESTED** [review]

**Review by DMotorygin — APPROVED** [review]

**Pull request merged**